cpp/demo/poisson_matrix_free/main.cpp

```diff
@@ -0,0 +1,215 @@
+// Copyright (C) 2022 Igor A. Baratta
+//
+// This file is part of DOLFINx (https://www.fenicsproject.org)
+//
+// SPDX-License-Identifier:    LGPL-3.0-or-later
+
+// =====================================
+// Matrix-free Conjugate Gradient solver
+// =====================================
+//
+// This demo illustrates how to:
+// * Solve a linear partial differential equation using a matrix free CG solver
```

Comment:
Is it worth mentioning here that this is not preconditioned and will therefore not scale with decreasing mesh size?

Reply:
I don't think we should mention explicitly what affects the convergence of the CG since the main aspect of the demo is the matrix-free operator, not the solver itself.

What about:
```c++
// .. note:: This demo illustrates the use of a matrix-free operator on an unpreconditioned CG solver, however for real-world applications using a preconditioner is highly advisable.
```

Reply:
Sounds good.

This demo illustrates the use of a matrix-free operator inside an unpreconditioned CG solver. A preconditioned CG solver should be used for real-world problems.